lib/backend/buffer_test.go

```diff
@@ -134,11 +137,69 @@ func (s *BufferSuite) TestWatcherSimple(c *check.C) {
 	}
 }
 
+// TestWatcherCapacity checks various watcher capacity scenarios
+func (s *BufferSuite) TestWatcherCapacity(c *check.C) {
+	const gracePeriod = time.Second
+
+	ctx := context.TODO()
+	b := NewCircularBuffer(
+		BufferCapacity(1),
+		BacklogGracePeriod(gracePeriod),
+	)
+	defer b.Close()
+	b.SetInit()
+
+	w, err := b.NewWatcher(ctx, Watch{
+		QueueSize: 1,
+	})
+	c.Assert(err, check.IsNil)
+	defer w.Close()
+
+	select {
+	case e := <-w.Events():
+		c.Assert(e.Type, check.Equals, types.OpInit)
+	case <-time.After(100 * time.Millisecond):
+		c.Fatalf("Timeout waiting for event.")
+	}
+
+	// emit and then consume 10 events.  this is much larger than our queu size,
+	// but should succeed since we consume within our grace period.
+	for i := 0; i < 10; i++ {
+		b.Emit(Event{Item: Item{Key: []byte{Separator}, ID: int64(i + 1)}})
+	}
+	for i := 0; i < 10; i++ {
+		select {
+		case e := <-w.Events():
+			c.Assert(e.Item.ID, check.Equals, int64(i+1))
+		case <-time.After(100 * time.Millisecond):
+			c.Fatalf("Timeout waiting for event.")
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		b.Emit(Event{Item: Item{Key: []byte{Separator}, ID: int64(i + 10)}})
+	}
+
+	// force grace period to expire
+	time.Sleep(gracePeriod)
```

Comment:
Could we use a fake clock and avoid the sleep? If you prefer not to we can also replace time.Now for testing.